pkg/services/authn/clients/ext_jwt.go

```diff
@@ -168,6 +168,20 @@ func (s *ExtendedJWT) authenticateAsService(accessTokenClaims authlib.Claims[aut
 		return nil, errExtJWTInvalidSubject.Errorf("unexpected identity: %s", accessTokenClaims.Subject)
 	}
 
+	permissions := accessTokenClaims.Rest.Permissions
+	fetchPermissionsParams := authn.FetchPermissionsParams{}
+	if len(permissions) > 0 {
+		fetchPermissionsParams.Roles = make([]string, 0, len(permissions))
+		fetchPermissionsParams.AllowedActions = make([]string, 0, len(permissions))
+		for i := range permissions {
+			if strings.HasPrefix(permissions[i], "fixed:") {
```

Comment:
So the idea here is the key remains the same in the jwt/CAP: `permissions`, but its content drives whether its considered a role or a permissions. That mostly works but I am working of 2 sets of deployments. HG Instances which are running a certain implementation based on the image they are running and then the MT services such query service which can upgrade to this later implementation out-of-band.

Won't using a separate key under CAP attributes and thereby a separate claim in the jwt be a better way of handling this?

If it was a separate key, I could update provisioned CAPs for the MT services to support both and perhaps `ext_jwt` could prefer one of the key above the other. Now, I know that its not optimal that you are not able to simply use `Permissions` when in fact you are now dealing with permissions (not roles), but that's more of a cosmetic problem. Wdyt?

Reply:
Maybe I am not understanding correctly, is `Permissions: []string{"fixed:folders:reader", "folders:read"},` a sure way to achieve backwards compatibility, if I were to modify my CAPs this way? That is, will the old implementation in rbac_sync just simply disregard non-roles? If that's so, then that makes me happy!

https://github.com/grafana/grafana/blob/main/pkg/services/authn/authnimpl/sync/rbac_sync.go#L46-L54

Reply:
From my current testing, a heterogeneous `permissions` value like below currently causes a 401 in ext_jwt as handled by Grafana.

```json
{
  "permissions": [
    "fixed:datasources:reader",
    "datasources:read",
    "datasources:query"
  ],
}

Not saying, that we can't first fix the current implementation to handle this heterogeneous value. If we do that, make sure that all the current HG deployments pick up the changes, then there might be a way to keep this PR the way it is, just reusing the `permissions` key.

The rollout becomes a 2-step process.

Reply:
On the other hand, I did get a chance to test this end to end with this branch and will comment on what I had to do to make it work.

Reply:
Yes with that approach we intend to allow using both roles and actions in the same list for backward compatibility.

Your 401s worry me though, I cannot explain them. Could it be that the ext_jwt client is not activated and you go through service account token authentication? Are there logs and traces we could look into? :pray:

Reply:
Edit: I found your traces:
```
logger=permissions.sync traceID=... t=2024-08-29T18:55:32.124331195Z level=error msg="Failed to fetch role from db" error="role not found" role=datasources:read
```
I think the issue is because we have the same error defined in both the enterprise and oss code base. So `errors.Is`  fails there:
https://github.com/grafana/grafana/blob/b0809c20003272c4b1640f26a539f370e39b8c67/pkg/services/authn/authnimpl/sync/rbac_sync.go#L87

Example: https://go.dev/play/p/MuYj2nZWo9N